sdk/core/azure-core/azure/core/polling/base_polling.py

```diff
@@ -0,0 +1,550 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+import abc
+import json
+from typing import TYPE_CHECKING, Optional, Any, Union
+
+from ..exceptions import HttpResponseError, DecodeError
+from . import PollingMethod
+
+if TYPE_CHECKING:
+    from azure.core.pipeline import PipelineResponse
+    from azure.core.pipeline.transport import (
+        HttpResponse,
+        AsyncHttpResponse,
+        HttpRequest,
+    )
+
+    ResponseType = Union[HttpResponse, AsyncHttpResponse]
+    PipelineResponseType = PipelineResponse[HttpRequest, ResponseType]
+
+
+try:
+    ABC = abc.ABC
+except AttributeError:  # Python 2.7, abc exists, but not ABC
+    ABC = abc.ABCMeta("ABC", (object,), {"__slots__": ()})  # type: ignore
+
+
+FINISHED = frozenset(["succeeded", "canceled", "failed"])
+FAILED = frozenset(["canceled", "failed"])
+SUCCEEDED = frozenset(["succeeded"])
+
+
+def finished(status):
+    if hasattr(status, "value"):
+        status = status.value
+    return str(status).lower() in FINISHED
+
+
+def failed(status):
+    if hasattr(status, "value"):
+        status = status.value
+    return str(status).lower() in FAILED
+
+
+def succeeded(status):
+    if hasattr(status, "value"):
+        status = status.value
+    return str(status).lower() in SUCCEEDED
+
+
+class BadStatus(Exception):
+    pass
+
+
+class BadResponse(Exception):
+    pass
+
+
+class OperationFailed(Exception):
+    pass
+
+
+def _as_json(response):
+    # type: (ResponseType) -> dict
+    """Assuming this is not empty, return the content as JSON.
+
+    Result/exceptions is not determined if you call this method without testing _is_empty.
+
+    :raises: DecodeError if response body contains invalid json data.
+    """
+    try:
+        return json.loads(response.text())
+    except ValueError:
+        raise DecodeError("Error occurred in deserializing the response body.")
+
+
+def _raise_if_bad_http_status_and_method(response):
+    # type: (ResponseType) -> None
+    """Check response status code is valid.
+
+    Must be 200, 201, 202, or 204.
+
+    :raises: BadStatus if invalid status.
+    """
+    code = response.status_code
+    if code in {200, 201, 202, 204}:
+        return
+    raise BadStatus(
+        "Invalid return status {!r} for {!r} operation".format(
+            code, response.request.method
+        )
+    )
+
+
+def _is_empty(response):
+    # type: (ResponseType) -> bool
+    """Check if response body contains meaningful content.
+
+    :rtype: bool
+    :raises: DecodeError if response body contains invalid json data.
+    """
+    content = response.text()
+    if not content:
+        return True
+    try:
+        return not json.loads(content)
```

Comment:
This is a somewhat expensive operation to just check if the response was a non-empty json payload, no?

Reply:
Maybe the name is misleading, but yes I want to check at the same if it's not empty, that it's valid JSON. suggestion?

Reply:
If I look at the point of usage(s), it seems like catching exceptions coming out of `_as_json` would be roughly equivalent. I'm not sure if there is ever a case where it "just being any kind of well formed json" is enough for total success - you usually require at least one key to be in the json as well, right?

If so, I'd get rid of this method and catch exceptions from _as_json.

Reply:
Ok, I simplify `_is_empty` to "is there bytes in this body"

Reply:
Pushed